nemo/collections/nlp/models/text_classification/ptune_text_classification_model.py

```diff
@@ -0,0 +1,461 @@
+from torch.nn.utils.rnn import pad_sequence
+
+from nemo.collections.nlp.data.text_classification.ptune_text_classification_dataset import (
+    PTuneTextClassificationDataset,
+    token_wrapper,
+)
+from nemo.collections.nlp.metrics.classification_report import ClassificationReport
+from nemo.collections.nlp.models.language_modeling.megatron_gpt_model import MegatronGPTModel
+from nemo.collections.nlp.models.nlp_model import NLPModel
+from nemo.collections.nlp.modules.common.megatron.megatron_init import initialize_model_parallel_for_nemo
+from nemo.collections.nlp.modules.common.prompt_encoder import PromptEncoder
+from nemo.collections.nlp.modules.common.tokenizer_utils import get_nmt_tokenizer
+from nemo.core.classes.common import typecheck
+from nemo.core.classes.exportable import Exportable
+from nemo.core.neural_types import LossType, NeuralType, PredictionsType, StringLabel, StringType
+from nemo.utils import logging
+
+__all__ = ['PTuneTextClassificationModel']
+
+
+SMALL_LOGITS = -100
+
+
+class PTuneTextClassificationModel(NLPModel, Exportable):
+    @property
+    def input_types(self) -> Optional[Dict[str, NeuralType]]:
+        return {"sentences": [NeuralType(('T'), StringType())], "labels": [NeuralType(('T'), StringLabel())]}
+
+    @property
+    def output_types(self) -> Optional[Dict[str, NeuralType]]:
+        return {
+            "floss": NeuralType((), LossType()),
+            "returned_pred": NeuralType(('B'), PredictionsType()),
+            "returned_label": NeuralType(('B'), PredictionsType()),
+        }
+
+    def __init__(self, cfg: DictConfig, trainer: Trainer = None):
+        """Initializes the PTune TextClassifier model."""
+        super().__init__(cfg=cfg, trainer=trainer)
+
+        initialize_model_parallel_for_nemo(
+            world_size=trainer.world_size,
+            global_rank=trainer.global_rank,
+            local_rank=trainer.local_rank,
+            tensor_model_parallel_size=cfg.get('tensor_model_parallel_size', 1),
+            seed=cfg.get('seed', 1234),
+        )
+
+        # shared params for dataset and data loaders
+        self.dataset_cfg = cfg.dataset
+        # tokenizer needs to get initialized before the super.__init__()
+        # as dataloaders and datasets need it to process the data
+        self.tokenizer = get_nmt_tokenizer(
+            library=cfg.tokenizer.library,
+            model_name=cfg.tokenizer.type,
+            tokenizer_model=self.register_artifact("tokenizer.model", cfg.tokenizer.model),
+            vocab_file=self.register_artifact("tokenizer.vocab_file", cfg.tokenizer.vocab_file),
+            merges_file=self.register_artifact("tokenizer.merges_file", cfg.tokenizer.merge_file),
+        )
+
+        self.class_weights = None
+
+        self.model = MegatronGPTModel.restore_from(
+            self.register_artifact('language_model.nemo_file', cfg.language_model.get('nemo_file', None)),
+            trainer=trainer,
+        ).half()
```

Comment:
Doesn't this depend on the config?

Reply:
added autocasting into the p-tune model.